### PR TITLE
chore: add bloom filter class

### PR DESCRIPTION
**Pull request by romange**
Based on https://github.com/jvirkki/libbloom implementation.

Unlike the original, our implementation uses XXH3 hash function to seed bit index generation. In addition, it assumes mi_malloc interface for dynamic allocation.

### CODE REVIEW

File: src/core/bloom.cc [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  uint64_t bits = uint64_t(ceil(entries * bpe));
+  bits = absl::bit_ceil(bits);  // make it power of 2.
+  if (bits < 1024) {
+    bits = 1024;
+  }
+
+  uint64_t length = bits / 8;
+
+  bf_ = (uint8_t*)mi_heap_calloc(heap, length, 1);
+
```
Comment (dranikpg) [inline src/core/bloom.cc:54]: just a thought, but an utuility that returns a unique_ptr on the mi-heap with a custom deleter would be helpful, we do this currently for some pointers in connection

File: src/core/bloom.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+   */
+  bool Add(std::string_view str);
+
+ private:
+  bool IsSet(size_t index) const;
+  bool Set(size_t index);  // return true if bit was set (i.e was 0 before)
+
+  uint64_t GetMask() const {
+    return (1ULL << bit_log_) - 1;
+  }
```
Comment (dranikpg) [inline src/core/bloom.h:36]: it's only called to pass it into BitIndex()

File: src/core/bloom.cc [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+inline bool Bloom::Set(size_t bit_idx) {
+  uint64_t byte_idx = bit_idx / 8;
+  bit_idx %= 8;
+
+  uint8_t b = bf_[byte_idx];
+  bf_[byte_idx] |= (1 << bit_idx);
+  return bf_[byte_idx] != b;
+}
```
Comment (dranikpg) [inline src/core/bloom.cc:103]: you actually don't need all of this if you just use `vector<bool>` 🤔

**Review by dranikpg — APPROVED** [review]
LGTM, but did you consider `vector<bool>`? Saves you half (most) of the code 🙂

File: src/core/bloom.cc (hunk 3 above)
Comment (romange) [reply src/core/bloom.cc:103]: You are right but then I need to pass mimalloc memory resource etc. it's not a lot of code and I won't object to reducing it in the future if memory usage won't increase.

File: src/core/bloom.h (hunk 2 above)
Comment (romange) [reply src/core/bloom.h:36]: Yes, you want to move it into cc?

File: src/core/bloom.cc (hunk 1 above)
Comment (romange) [reply src/core/bloom.cc:54]: Maybe but then we will increase the class size

**Review by dranikpg — APPROVED** [review]

**Pull request merged**